### ISSUE TITLE
Emtpy files detection is not working properly for hardlinks

### ISSUE THREAD

**Issue by hroncok** [rpminspect/rpminspect#1252]
**Describe the bug**
In the Zuul CI from https://src.fedoraproject.org/rpms/python3.12/pull-request/76 I got:

```
/usr/lib64/python3.12/test/test_capi/__pycache__/test_exceptions.cpython-312.opt-2.pyc became a non-empty file on x86_64

    Remedy: A previously empty file is no longer empty. Make sure this change is intended and fix the package spec file if necessary.

/usr/lib64/python3.12/test/__pycache__/test_xml_etree_c.cpython-312.opt-2.pyc became a non-empty file on x86_64

    Remedy: A previously empty file is no longer empty. Make sure this change is intended and fix the package spec file if necessary.

/usr/lib64/python3.12/test/test_sqlite3/__pycache__/test_factory.cpython-312.opt-2.pyc became an empty file on x86_64

    Remedy: A previously non-empty file is now empty. Make sure this change is intended and fix the package space file if necessary.
```

This does not seem correct. The files were not empty are are still not empty.

Perhaps it has to do with the fact that the files are sometimes hardlinked to others.

**To Reproduce**
Include the rpminspect command you ran including the NVRs of the builds passed to rpminspect.  Do not include a link to a Fedora, CentOS, or RHEL system integrating rpminspect.  If you are reporting a bug based on one of these, go to the *diagnostics* section of the rpminspect output and copy the rpminspect command string that was run.

```
  "diagnostics": [
    {
      "result": "DIAGNOSTICS",
      "message": "Version information for libraries and programs used by rpminspect as well as storage requirements.  This result is for informational and diagnostic purposes only.",
      "details": "rpminspect version 1.12\nzlib version 1.2.13\nlibmagic version 544\nlibclamav version 1.0.2\nlibrpm version 4.18.1\nlibxml version 21004\njson-c version 0.16\nlibcurl version 8.0.1 (OpenSSL\/3.0.8 zlib\/1.2.13 brotli\/1.0.9 libidn2\/2.3.4 libpsl\/0.21.2 (+libidn2\/2.3.4) libssh\/0.10.5\/openssl\/zlib nghttp2\/1.52.0)\nlibarchive version 3.6.1 (zlib\/1.2.13 liblzma\/5.2.9 bz2lib\/1.0.8 liblz4\/1.9.4 libzstd\/1.5.2)\nlibyaml version 0.2.5\nOpenSSL version 3.0.8 7 Feb 2023\nxmlrpc-c version 1.51.8\nmsgunfmt (GNU gettext-tools) 0.21.1\nannocheck version 12.23.\nabidiff version 2.2.0\nkmidiff version 2.2.0\nudevadm version 253\n\nSpace required to download artifacts: 0 bytes (0 B)\nSpace required to unpack artifacts: 519353678 bytes (495 MiB)\n"
    },
    {
      "result": "DIAGNOSTICS",
      "message": "Command line arguments used to invoke rpminspect.",
      "details": "\/usr\/bin\/rpminspect -c \/usr\/share\/rpminspect\/fedora.yaml -Dv -k -t BAD -w \/var\/tmp\/rpminspect -a x86_64,noarch,src -o \/root\/zuul-output\/logs\/result.json -F json \/var\/tmp\/rpminspect\/python3.12-3.12.0~rc1-1.fc39 \/var\/tmp\/rpminspect\/python3.12-3.12.0~rc2-1.fc39"
    },
    {
      "result": "DIAGNOSTICS",
      "message": "Local configuration file: rpminspect.yaml",
      "details": "# exclude test XML data (not always valid) from XML validity check:\nxml:\n    ignore:\n        - \/usr\/lib*\/python*\/test\/xmltestdata\/*\n        - \/usr\/lib*\/python*\/test\/xmltestdata\/*\/*\n# exclude _socket from ipv4 only functions check, it has both ipv4 and ipv6 only\nbadfuncs:\n    allowed:\n        \/usr\/lib*\/python*\/lib-dynload\/_socket.*:\n            - inet_aton\n            - inet_ntoa\n# exclude the debug build from annocheck entirely\nannocheck:\n    ignore:\n        - \/usr\/bin\/python*d\n        - \/usr\/lib*\/libpython*d.so.1.0\n        - \/usr\/lib*\/python*\/lib-dynload\/*.cpython-*d-*-*-*.so\n# don't report changed content of compiled files\n# that is expected with every toolchain update and not reproducible yet\nchangedfiles:\n    # note that this is a posix regex, so no \\d\n    exclude_path: (\\.so(\\.[0-9]+(\\.[0-9]+)?)?$|^\/usr\/bin\/python[0-9]+\\.[0-9]+d?m?$)\n# files change size all the time, we don't need to VERIFY it\n# however, the INFO is useful, so we don't disable the check entirely\nfilesize:\n    # artificially large number, TODO a better way\n    size_threshold: 100000\n# completely disabled inspections:\ninspections:\n    # we know about our patches, no need to report anything\n    patches: off"
    },
    {
      "result": "DIAGNOSTICS",
      "message": "Before Build",
      "details": "\/var\/tmp\/rpminspect\/python3.12-3.12.0~rc1-1.fc39"
    },
    {
      "result": "DIAGNOSTICS",
      "message": "After Build",
      "details": "\/var\/tmp\/rpminspect\/python3.12-3.12.0~rc2-1.fc39"
    }
  ],
```

**Expected behavior**
Files that are not empty should not be reported as empty.

**Environment:**
 - Zuul CI

**Comment by dcantrell** [comment]
I can't find python3.12-3.12.0~rc2-1.fc39 in Koji.

**Comment by hroncok** [comment]
It's from a pull request I linked.

**Comment by dcantrell** [comment]
Oh, derp.  Heh.

**Comment by hroncok** [comment]
https://koji.fedoraproject.org/koji/taskinfo?taskID=105824060

**Comment by dcantrell** [comment]
OK, so I got that scratch build and I am trying to reproduce the problem locally and I can't.  I see that it fails in Zuul.  The empty file error comes from the filesize inspection, but as you say those files are not empty (nor were they in the previous build).

To me this is probably a Zuul environment issue.  Are you able to reproduce the issue locally?

@morucci could use some Zuul assistance here.  I do not know how Zuul is setting up the environment in which to run rpminspect.

**Comment by morucci** [comment]
@dcantrell, sure
- rpminspect is run on this container image: quay.io/software-factory/zuul-worker-f38
- here is the job logs https://fedora.softwarefactory-project.io/logs/76/76/614cafc40073fcdd4df1543d966efe651d7de3ad/check/rpm-rpminspect/5e456bc/job-output.txt

**Comment by hroncok** [comment]
>  Are you able to reproduce the issue locally?

Yes, I am:

```
$ tree
.
├── python3.12-3.12.0~rc1-1.fc39
│   ├── python-unversioned-command-3.12.0~rc1-1.fc39.noarch.rpm
│   ├── python3-debug-3.12.0~rc1-1.fc39.x86_64.rpm
│   ├── python3-devel-3.12.0~rc1-1.fc39.x86_64.rpm
│   ├── python3-idle-3.12.0~rc1-1.fc39.x86_64.rpm
│   ├── python3-libs-3.12.0~rc1-1.fc39.x86_64.rpm
│   ├── python3-test-3.12.0~rc1-1.fc39.x86_64.rpm
│   ├── python3-tkinter-3.12.0~rc1-1.fc39.x86_64.rpm
│   └── python3-3.12.0~rc1-1.fc39.x86_64.rpm
├── python3.12-3.12.0~rc2-1.fc39
│   ├── python-unversioned-command-3.12.0~rc2-1.fc39.noarch.rpm
│   ├── python3-debug-3.12.0~rc2-1.fc39.x86_64.rpm
│   ├── python3-devel-3.12.0~rc2-1.fc39.x86_64.rpm
│   ├── python3-idle-3.12.0~rc2-1.fc39.x86_64.rpm
│   ├── python3-libs-3.12.0~rc2-1.fc39.x86_64.rpm
│   ├── python3-test-3.12.0~rc2-1.fc39.x86_64.rpm
│   ├── python3-tkinter-3.12.0~rc2-1.fc39.x86_64.rpm
│   └── python3-3.12.0~rc2-1.fc39.x86_64.rpm
└── workdir

4 directories, 16 files

$ rpminspect -c /usr/share/rpminspect/fedora.yaml -Dv -k -t BAD -w workdir -a x86_64,noarch,src -o result.json -F json python3.12-3.12.0~rc1-1.fc39 python3.12-3.12.0~rc2-1.fc39
...
Running files inspection...          FAIL
...

$ grep 'empty file' result.json 
      "message": "\/usr\/lib64\/python3.12\/test\/test_capi\/__pycache__\/test_exceptions.cpython-312.opt-2.pyc became a non-empty file on x86_64"
      "message": "\/usr\/lib64\/python3.12\/test\/__pycache__\/test_xml_etree_c.cpython-312.opt-2.pyc became a non-empty file on x86_64"
      "message": "\/usr\/lib64\/python3.12\/test\/test_sqlite3\/__pycache__\/test_factory.cpython-312.opt-2.pyc became an empty file on x86_64"


$ rpminspect --version
rpminspect version 1.12

$ rpm -qa | grep rpminspect
librpminspect-1.12-0.1.202308301636git23a98a0.fc37.x86_64
rpminspect-1.12-0.1.202308301636git23a98a0.fc37.x86_64
rpminspect-data-fedora-1.11-0.1.202308281354gitcda52d4.fc37.noarch
```

**Comment by dcantrell** [comment]
I think this one came down to updating the path globs in the global ignore list in fedora.yaml in rpminspect-data-fedora.  I did that here:

https://github.com/rpminspect/rpminspect-data-fedora/pull/45

@hroncok Can you try that locally and see if it resolves what you were seeing?

**Comment by hroncok** [comment]
I don't undedstand why such files need to be excluded. Isn't that just covering up a broken detection when hardlinks are involved?

I am unable to test it now, but the pycache dirs can be in subdirectories, does the glob work that way?

**Comment by dcantrell** [comment]
> I don't undedstand why such files need to be excluded. Isn't that just covering up a broken detection when hardlinks are involved?

Probably.  I noticed the global ignores were not catching all of the __pycache__ files.  I remember some time ago adding those to the ignore list because they change with every build so it doesn't make sense to always check them.  But that may actually be covering up bad size reading here.

Taking a further look, but I think that the global ignores in rpminspect-data-fedora should still be updated.  Unless you think we should drop those entirely.

> I am unable to test it now, but the pycache dirs can be in subdirectories, does the glob work that way?

Yes, it should work.

**Comment by hroncok** [comment]
I think they should not be excluded from this check.

**Comment by dcantrell** [comment]
So this appears to come down to extracting the RPM payload and reading the file size from the header.  For hard links, the size is zero but the link count is > 1.  This makes sense for the purposes of the archive.  What I've done is extend the filesize inspection so in cases where the file is a hard link and the size is zero, stat(2) the one on disk and save that size to the rpmfile_entry_t struct.  It could still be zero, but it will at least read the size of the file as expected.

I thought about skipping the inspection for hard links, but that seems like a workaround.  Hard link names could change and that can be worth seeing in the results.

Sadly, there's not an easy way to get the size from the payload header either.  It would be another loop because it could link to an as-yet unextracted archive entry.  While I don't like adding additional stat(2) calls, this method is sort of straightforward.

This patch gets rid of the "became a non-empty" or "became empty" notices for hard links.  Those are from the filesize inspection.

PR coming.
